### ISSUE TITLE
Issue with new MoneroOcean XMRig-mo implementation in 4.7.9.8 (with workaround)

### ISSUE THREAD

**Issue by coolbreeze21** [RainbowMiner/RainbowMiner#1928]
Thank you for the effort to get the XMRig-mo fork working, which was necessary to mine RTM on MO.  It can now do this successfully. But I think there is an issue with it in RBM 4.7.9.8. 

The first several launches (at least) of XMRig-mo are a bumpy road.  It appears that the MO fork wants to briefly benchmark all the algos it supports (@ 20 seconds each) and write the results to the XMRig-mo config.json, so that it can use it for intelligent algo switching.  It seems that when RBM launches XMRig-mo it doesn’t allow it to complete this benchmarking.  So RBM thinks XMRig-mo mined little or nothing, because it was benchmarking, and switches away.  Since XMRig-mo thinks benchmarking didn’t complete, it restarts the process the next time it is launched.  This wastes several minutes at the start of every launch.  Without user intervention, the only way this is overcome is if XMRig-mo is lucky enough to get an extended period (5-10 minutes) to complete benchmarking and start actually mining RTM.

I tried overcoming this behavior by adding

"rebench-algo": false,

to the json.  This was unsuccessful.

I then tried overcoming it by adding an entire benchmark results block from an older XMRig json I had as follows (result numbers are specific to the rig in question):

"rebench-algo": false,
    "bench-algo-time": 20,
    "algo-min-time": 0,
    "algo-perf": {
        "cn/0": 554.5055863005116,
        "cn/1": 575.4207355079992,
        "cn/2": 575.4207355079992,
        "cn/r": 575.4207355079992,
        "cn/fast": 1150.8414710159984,
        "cn/half": 1150.8414710159984,
        "cn/xao": 575.4207355079992,
        "cn/rto": 575.4207355079992,
        "cn/rwz": 767.2276473439989,
        "cn/zls": 767.2276473439989,
        "cn/double": 287.7103677539996,
        "cn/ccx": 1109.0111726010232,
        "cn-lite/0": 2373.278995484084,
        "cn-lite/1": 2373.278995484084,
        "cn-heavy/xhv": 256.2979189485213,
        "cn-pico": 14539.971171970285,
        "cn-pico/tlo": 14539.971171970285,
        "cn/gpu": 149.2046077893582,
        "rx/0": 5394.687106257876,
        "rx/arq": 38562.31578947369,
        "rx/graft": 5276.00336205085,
        "rx/sfx": 5394.687106257876,
        "panthera": 2512.6050420168067,
        "argon2/chukwav2": 11927.917499736925,
        "astrobwt": 1161.0822060353798,
        "kawpow": 6.013470016991623e-154,
        "ghostrider": 1250
    },

This was successful on 2 different rigs and caused XMRig-mo to go straight to mining, without benchmarking, when launched by RBM.

I did not test it, but it strikes me that this benchmarking issue might occur with XMRig-mo and other coins besides RTM.

Both before and after my json modification, I think I noticed that the RBM dashboard occasionally stops updating the XMRig-mo A/R/S stats, even though the console shows XMRig-mo is still mining successfully.  When this happens, it is just a matter of time until RBM unnecessarily shuts down XMRig-mo.

I think it must have been an impressive piece of work to integrate the XMRig-mo fork with RBM’s existing XMRig implementation.  However, I think it might be better to just let XMRig-mo have its own separate identity as the miner that should be used with MO, rather than trying to integrate it with the main XMRig branch.  XMRig-mo is updated nearly as frequently as XMRig itself, clearly has its own peculiarities, and has 0% dev fee when used on MO.  Just a thought.

**Comment by RainbowMiner** [comment]
Thank you very much for your feedback! I will think through the alternatives

**Comment by coolbreeze21** [comment]
You are welcome.  Thank you again for your work in getting this mostly working in the first place.  RTM has been one of the most profitable coins to CPU mine for several months, but I have enough RTM already and would rather mine RTM and get paid in XMR on MO.  Also, I think I need to make a correction: I think XMRig-mo _**can be set**_ to 0% dev fee if used on MO, unlike the main XMRig miner, but the default setting may be higher than 0%, so a quick and easy json edit is necessary to do this.

**Comment by coolbreeze21** [comment]
Also, I realized that XMRig-mo's default behavior, and your work in getting it to target a specific coin, allows for the possibility of both a "MoneroOcean" option (where MO decides what coin to mine) and a "MoneroOceanCoins" option (where RBM decides what coin to mine).  They could presumably use the same one-time "algo-perf" benchmarking results json block, but XMRig-mo needs to be initially given 5-10 minutes of benchmarking time and the ability to write the "algo-perf" results and reset the "rebench-algo" flag from true to false.

By the way, the same behavior applies to C3Pool (https://c3pool.com/oldui/en/), which is nearly identical to MoneroOcean except for 75% lower transaction fees (0.0001 vs. 0.0004 XMR per payout).  Both pools have no pool fees, except for the payout fees, and both mine the same coins.  On some coins one has a higher hashrate and on some coins the other does.  I would welcome RBM adding C3Pool to its pools list.  I don't think the C3Pool XMRig fork uses MO's unusual stratum over rpc2 (at least I recall using vanilla XMRig on C3Pool.

The "algo-perf" results blocks I used, and my knowledge of the miner's behavior, was taken from my experience for several months with C3Pool.  I liked them a lot, though there isn't much support if you don't speak Chinese.  They are Singapore based, but have servers in the EU (3), Asia (3) and US (1).  But If you know the MO interface, the C3Pool "old" interface (which still works) is identical.  And they do get pool fees, including payouts, down to almost zero (.0001 XMR = 2 cents) with minimum payouts as low as .003 XMR = $0.60!

**Comment by RainbowMiner** [comment]
It's better to combine the xmrig and xmrig-mo into one module. Splitting would imply duplicate benchmarking and inconsistent selection with MRR.

I'll have a look into  C3Pool, now. This pool sounds promising, I'll add it.

**Comment by coolbreeze21** [comment]
I was hoping that C3Pool would be added to the pools in 4.8.0.0, but it looks like no such luck.  Or was it added, but I messed things up by trying (and failing) to add C3Pool to my setup as a custom pool?  Or did adding it turn out to be too much of pain in the ass to implement?

**Comment by RainbowMiner** [comment]
I just hadn't the time to implement the module, yet. To be expected within the next 2-3 days. Sorry, that you have to wait.

**Comment by coolbreeze21** [comment]
I don't mind the wait, I was just hoping you hadn't abandoned doing this.  So, I was very happy with your response.  As a reminder, C3Pool uses its own fork of XMRig, just like MoneroOcean does.  The C3Pool fork is here: https://github.com/C3Pool/xmrig-C3/releases.  The fork is probably why my attempts to set up C3Pool as a custom pool were failures.  Hopefully, it is not too much of a pain for you to integrate xmrig-C3 while the steps you took to integrate xmrig-mo are still relatively fresh in your mind.  The two pools appear almost identical in their user interfaces.  I think it is worth the effort because of C3Pool's payout fees being 75% less than MoneroOcean's.  Thanks a lot!

**Comment by RainbowMiner** [comment]
Ok, added. I have adapted the `./Miners/Xmrig.ps1` module as well, so that the xmrig-mo is being used for RTM mining on C3pool, too. I have found no differences in the source code.

**Comment by coolbreeze21** [comment]
Great!  I am glad that their fork didn't add any extra complications.  Thanks a lot!  Do I need to wait for the next update to test and use this?

**Comment by RainbowMiner** [comment]
No, you can just as well download the current master.zip and extract it into your RainbowMiner folder. Alternatively, just download the following files directly into their destination folders (use right-click and save-as):
- into folder `./Pools`: [./Pools/C3pool.ps1](https://raw.githubusercontent.com/RainbowMiner/RainbowMiner/master/Pools/C3pool.ps1)
- into folder `./Balances`: [./Balances/C3pool.ps1](https://raw.githubusercontent.com/RainbowMiner/RainbowMiner/master/Balances/C3pool.ps1)
- into folder `./Data`: [./Data/PoolsConfigDefault.ps1](https://raw.githubusercontent.com/RainbowMiner/RainbowMiner/master/Data/PoolsConfigDefault.ps1)
- into folder `./Miners`: [./Miners/Xmrig.ps1](https://raw.githubusercontent.com/RainbowMiner/RainbowMiner/master/Miners/Xmrig.ps1)

.. then let RainbowMiner run for a couple of rounds and the C3pool record should be created in pools.config.txt. 
Thank you for testing this new module. I am particularly interested, if the Ghostrider/Take2 mining (RTM coin) works on the stratum address mine.c3pool.com, or if we have to change it to rtm.c3pool.com. If not, please let me know.

**Comment by coolbreeze21** [comment]
I used the second, surgical approach and only added/replaced the four files.  RBM immediately wanted to CPU mine RTM on C3Pool instead of MO, which it was currently mining RTM on (I hadn't disabled MO to force it).  However, it wanted to mine it using CpuminerWyvern, which is faster than XMRig on this 2080 rig (not on my 3090) because it has a custom tune_config file.  This failed with the error messages "Stratum authentication failed" and "Stratum answer id is not correct!".  I then turned off CpuminerWyvern for Take2 and RBM switched to XMRig as its second choice.  This worked successfully.  Everything looked good on the console and 50+ valid shares were seen on the C3Pool side.

So, I think the stratum address is good, but it may be necessary to do more to force the use of XMRig only on C3Pool, as with MO.

**Comment by RainbowMiner** [comment]
Ah, sorry, I just noticed, that I have already disabled the MoneroOcean + RTM for all miners, except that xmrig-mo. I'll add C3pool to the other miner's exclude lists.

**Comment by coolbreeze21** [comment]
When I followed the same procedure on the 5950X/3090 rig, RBM immediately wanted to benchmark RandomX on C3Pool on both the CPU (with XMRig and SrbMinerMulti) and the 3090 (with XMRig).  All three tests appeared successful on the pool side (SrbMinerMulti isn't visible in the RBM console).  It wasn't necessary to disable CpuminerWyvern on the 5950X, because XMRig was faster than Wyvern on this rig, even when fed its own custom tune_config file.  This is consistent with the XMRig devs' assertion that their superiority over tuned Wyvern increases with better CPUs.  After benchmarking completed, this rig settled into CPU mining RTM using XMRig on C3Pool and went back to its usual GPU mining ETH + ZIL using Phoenix on Rustpool.  So, it looks like SrbMinerMulti will work with C3Pool for RandomX (it sucks at Take2 in past benchmarking, 1.17 KH/s, 4% rejected vs. XMRig 4.43 KH/s, 0% rejected).  Also, C3Pool is now visible in my stats with a climbing balance.

I am a very happy camper.  Thank you so much!

**Comment by coolbreeze21** [comment]
Incidentally, I think it is interesting that your getting RTM mining working on MO recently coincided with MO soaring from a negligible #20 to #2 among RTM pools right now.  It did this with a mere 1300 miners (#3 and #4 have 6000).  Coincidence?  I wonder.  C3Pool is currently #5 with only 230 miners.  It will be interesting to see what happens in a few more weeks once it is widely available on RBM and users have had a chance to consider enabling it.

**Comment by coolbreeze21** [comment]
A couple of updates:

1. I moved on to testing a SSL connection to C3Pool on both rigs.  It worked.  Nice job on making the SSL option available and getting the configuration correct.

2. While testing this, the i9-9900K/2080 rig decided to switch RTM mining from XMRig to SrbMinerMulti.  This may have been because the pool hit a string of tough ghostrider rotations, dragging the XMRig hashrate down.  In any case, SrbMinerMulti RTM mining failed on C3Pool with errors "algo array must include at least one supported pool algo: [ghostrider] [ghostrider]" and "PARSE error: 'params' is not an array in received JSON [ghostrider]".  So, I think your decision to exclude all miners but XMRig-mo is the right one.

3. I noticed that there seems to be a glitch in setting the payout threshold on C3Pool.  It doesn't seem to work using the old user interface.  It does work using the new user interface.  I think I had noticed this months ago.  This is a pool issue, not a RBM issue.  However, to avoid unnecessary user frustration, I suggest changing the dropdown details text for C3Pool in the RBM help file to something like "Old User Interface https://c3pool.com/oldui/en/ will be familiar to anyone who has used MoneroOcean. New UI https://c3pool.com/.  Auto-exchange and payout in XMR, enter your Monero wallet address and optionally a password (either a real password or your email-address) into pools.config.txt to start mining.  It may be necessary to use the new UI to set payout threshold.".  C3Pool doesn't require a password (or anything) to change the payout threshold and I never entered anything into my RBM password settings when testing, just left it at the default.
